client/src/nonblocking/quic_client.rs

```diff
@@ -139,21 +143,29 @@ impl QuicNewConnection {
 
         let connecting = endpoint.connect(addr, "connect").unwrap();
         stats.total_connections.fetch_add(1, Ordering::Relaxed);
-        let connecting_result = connecting.await;
-        if connecting_result.is_err() {
-            stats.connection_errors.fetch_add(1, Ordering::Relaxed);
-        }
-        make_connection_measure.stop();
-        stats
-            .make_connection_ms
-            .fetch_add(make_connection_measure.as_ms(), Ordering::Relaxed);
+        if let Ok(connecting_result) = timeout(
+            Duration::from_millis(QUIC_CONNECTION_HANDSHAKE_TIMEOUT_MS),
+            connecting,
+        )
+        .await
+        {
+            if connecting_result.is_err() {
+                stats.connection_errors.fetch_add(1, Ordering::Relaxed);
+            }
+            make_connection_measure.stop();
+            stats
+                .make_connection_ms
+                .fetch_add(make_connection_measure.as_ms(), Ordering::Relaxed);
 
-        let connection = connecting_result?;
+            let connection = connecting_result?;
 
-        Ok(Self {
-            endpoint,
-            connection: Arc::new(connection),
-        })
+            Ok(Self {
+                endpoint,
+                connection: Arc::new(connection),
+            })
+        } else {
+            Err(ConnectionError::TimedOut.into())
```

Comment:
let's `impl From<tokio::time::error::Elapsed> for ConnectionError` and `?` the timeout error away instead of taking the indent

Reply:
Rust apparently doesn't like it when I try to impl traits where all the types are not defined in the current crate. Incidentally, I actually do have another PR that would replace the WriteError with a locally-defined error type, which would allow us to impl From. So I think it's better to leave this for a followup.